### ISSUE TITLE
functionality to rename a project

### ISSUE THREAD

**Issue by ewdurbin** [pypa/warehouse#1919]
At this point, there are some admin tools in pypa/pypi-legacy i'm not comfortable using to rename projects when users request such an operation. Some examples are modifications of case or underscore/hyphen swaps which are equivalent after applying [`safe_name`](https://github.com/pypa/setuptools/blob/d6ffbbd869548f855e7550cf28686e88b48a1495/pkg_resources/__init__.py#L1373-L1378) (`myPackage` => `mypackage`, or `my-package` -> `my_package`)
 
We should establish a process and enforce it via whatever mechanism we use to perform such an operation (code, admin ui, carrier pigeon).

**Comment by dstufft** [comment]
Renaming doesn't really work in general because even if you rename the file, the contents of said file are going to still be using the old name. The only renames that really work at all are if they normalize to the same thing.

**Comment by dstufft** [comment]
Well, I guess we could allow you to rename as long as the normalization still matched, would that satisfy this issue?

**Comment by brainwane** [comment]
In our [meeting today](https://wiki.python.org/psf/PackagingWG/2017-12-19-Warehouse) we double-checked and agreed that this new admin feature can wait till after Warehouse launch.

**Comment by jacebrowning** [comment]
(I was brought here by https://twitter.com/brainwane/status/976489373607772161)

Here's my use case:

https://pypi.python.org/pypi/coverage-space was originally named `coverage.space`, but I realized this was problematic because that name requires quoting to include in a `Pipfile`. I thought I could upload newer versions of the PyPI package as `coverage-space`, but I was getting an error message along the lines of "name must start with coverage.space". I couldn't find a way to rename the PyPI package so I ended up deleting all versions and uploading 1.0 as `coverage-space`.

**Comment by brainwane** [comment]
Thanks @jacebrowning and I'm sorry for the trouble! We'll take this into account as we work on Warehouse and improve documentation and user support processes.

`coverage.space`/`coverage-space` looks like a cool project! ([Here it is in the new PyPI](https://pypi.org/project/coverage-space/).)

**Comment by jacebrowning** [comment]
Is there any way I can upload a project under both names (or set up a redirect) to avoid broken downloads?

**Comment by ewdurbin** [comment]
the two project names share the same normalized name, so a rename would have been more tenable before project deletion. unfortunately we don’t have tooling for a rename, so that would have had to have been built.

as far as what we do now, restoring deleted projects may be doable but is not a task we’ve undertaken before.

filenames cannot be reused, and I believe that will remain the strict rule, so the only real recourse to a project deletion is to upload replacement `post0` or similar releases with the new name.

**Comment by jacebrowning** [comment]
Thanks @ewdurbin. Based on a rough search, I don't see more than 1000 projects depending on the package I "renamed" and probably less than a few percent are actively maintained. I think I can manage to support those few affected.

It might be nice if going forward "normalized" names are treated as equivalent. That is to say:
- An admin of "foo-bar" could upload "foo.bar" to the same package
- A non-admin of "foo-bar" could never upload "foo.bar"

**Comment by di** [comment]
Blocked on https://github.com/pypa/warehouse/issues/4772.

**Comment by di** [comment]
No longer blocked.

**Comment by brainwane** [comment]
@di do you have any policy/architecture concerns about this, or is it open for someone to implement? Is this something we should seek volunteer help to work on?

Since this potentially blocks #1506 I figure it's something worth pushing on.

**Comment by dstufft** [comment]
I think this works now? If you upload a project and change the name, as long as it normalizes to the same name as before, the project will be renamed on PyPI. We don't support arbitrary renaming where you can rename foo to bar, but I don't think we ever really could?

**Comment by di** [comment]
Yeah, since #4783 we have the ability to rename a project when a new release is made.

This issue is a bit old, I think it might intend to allow admins or owners to rename a project _without_ making a new release, but I'm not really sure that's necessary.

**Comment by brainwane** [comment]
@ewdurbin and @jacebrowning could you reply and let us know whether your use cases are now taken care of?

**Comment by ewdurbin** [comment]
looks like the "different name same normalization" use-case is covered by #4783 to me!

**Comment by jacebrowning** [comment]
Yes, I believe my use cases are now taken care of.

**Comment by brainwane** [comment]
OK, great, glad that's cleared up and sorry for misunderstanding!
